contentcuration/contentcuration/static/js/edit_channel/models.js

```diff
@@ -1,63 +1,284 @@
 var Backbone = require("backbone");
 var _= require("underscore");
 
-window.ChannelEditDataModel = Backbone.Model.extend({
-	urlRoot: '/channel_edit',
-	initialize: function() {
-		//_.bindAll(this);
-	}
-});
-
-/* TODO: CHANGE TO ChannelEditDataModel.EXTEND */
-var TopicNodeModel = Backbone.Model.extend({
+var NodeModel = Backbone.Model.extend({
 	defaults: {
-        color1:"",
-		color2:"",
-		color3:"",
 		title:"Untitled",
-		description:"This is the description generated by default"
-    }
+		description:"No description",
+		parent: null,
+		children:[],
+    },
+    urlRoot: function() {
+		return window.Urls["node-list"]();
+	},
+	toJSON: function() {
+	  var json = Backbone.Model.prototype.toJSON.apply(this, arguments);
+	  json.cid = this.cid;
+	  return json;
+	},
+	parse: function(response) {
+	    return _(response).isArray()
+	         ? response[0]
+	         : response;
+	},
+
+	/*Used when copying items to clipboard*/
+    duplicate: function(parent_id){
+    	var title = this.generate_title(this.get("title"));
+		var data = {
+			title: title,
+			created : this.get("created"),
+			modified : this.get("modified"),
+			description: this.get("description"),
+			deleted: this.get("deleted"),
+			sort_order : this.get("sort_order"),
+			license_owner : this.get("license_owner"),
+			license: this.get("license"),
+			kind: this.get("kind"),
+			parent: parent_id
+		};
+		var node_data = new NodeModel(data);
+		var self = this;
+		node_data.save(data, {async:false,
+			success:function(){
+				self.copy_children(node_data, self.get("children"));
+			}
+		});
+		return node_data;
+	},
+
+	/* Function in case want to append (Copy #) to end of copied content*/
+	generate_title:function(title){
+    	var list = this.attributes.title.split(" ");
+    	if(list[list.length - 1] == "(Copy)"){ 				//model has been copied once before
+    		list[list.length - 1] = "(Copy 2)";
+    		title = list.join(" ");
+    	}else if(list.length > 2 							//model has been copied multiple times
+    			&& list[list.length-2] == "(Copy" 
+    			&& list[list.length-1].includes(")")){
+    		var copy_number = list[list.length-1].replace(")","");
+    		list[list.length-1] = ++copy_number + ")";
+			title = list.join(" ");
+    	}else{
+    		title += " (Copy)";
+    	}
+	},
+	copy_children:function(node, original_collection){
+		var self = this;
+		var parent_id = node.id;
+		var copied_collection = new NodeCollection();
+		copied_collection.get_all_fetch(original_collection);
+		$(copied_collection.models).each(function(){
+			console.log("end",this.duplicate(parent_id));
+		});
+	},
+	validate:function (attrs, options){
```

Comment:
I know we had discussed this previously, but I am wondering if it might be more efficient to do these validation checks on the server side so that we just do a single API call that does all the validation for us, rather than having to fetch both parents and siblings and doing all these checks.

Also, on the server we have Django-mptt which will be able to much more efficiently check that the tree is not being manipulated in a bad way.